src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -508,21 +606,43 @@ protected virtual void ValidateCHash(OpenIdConnectProtocolValidationContext vali
                 return;
             }
 
-            object cHashClaim;
-            if (!validationContext.ValidatedIdToken.Payload.TryGetValue(JwtRegisteredClaimNames.CHash, out cHashClaim))
-            {
-                throw LogHelper.LogExceptionMessage(new OpenIdConnectProtocolInvalidCHashException(LogHelper.FormatInvariant(LogMessages.IDX21307, validationContext.ValidatedIdToken)));
-            }
+            string alg;
+            string chash;
 
-            var chash = cHashClaim as string;
-            if (chash == null)
+            if (validationContext.ValidatedIdTokenJwt != null)
```

Comment:
It may be cleaner to have this method transform the representation of the JWT into a normalized form then call an internal method that has a single format.